### PR TITLE
Add migrations for HASH Graph 

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This change adds migration scripts for the new type system's DB schemas. 
Link types are not a part of this yet, and some of the aspects of the schemas are best guesses, and are subject to change.

## 🔗 Related links

- [Asana task 1](https://app.asana.com/0/1201095311341924/1202510174412966/f) _(internal)_
- [Asana task 2](https://app.asana.com/0/1202482456346912/1202510174412964/f) _(internal)_

## 🔍 What does this change?

- Moves previous `packages/hash/datastore` package to `packages/graph/migrations`
- Changes DB schema to match new the type system
- Very initial Rust types for the DB 
- Change to `.env` so we do not wipe people's local DBs when trying out the new Graph API

## 📜 Does this require a change to the docs?

- We did change some of the documentation for the `migrations` project, although we need to add more 'getting started' documentation to the new Graph API

## 🐾 Next steps

- Merging https://github.com/hashintel/hash/pull/710
- Considering if the .env variable used by SQLX can be aligned with what we have in the root .env

## 🛡 What tests cover this?

- No automated tests yet

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  run `yarn pg:recreate-db` and `yarn db:migrate up` from the migrations directory
1.  Confirm that the DB is created

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -12,13 +12,13 @@ HASH_OPENSEARCH_PASSWORD=admin
...
 
-HASH_PG_DATABASE=postgres
+HASH_PG_DATABASE=postgres_graph
 HASH_PG_HOST=localhost
 HASH_PG_PASSWORD=postgres
 HASH_PG_PORT=5432
 HASH_PG_USER=postgres
 
-HASH_MIGRATION_PG_URL=postgres://postgres:postgres@localhost/postgres
+HASH_MIGRATION_PG_URL=postgres://postgres:postgres@localhost/postgres_graph
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply .env:21]: This is fine

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    "ids",
+    {
+      version_id: {
+        type: "UUID",
+        primaryKey: true,
+      },
+      id: {
+        type: "UUID",
+        notNull: true,
+        unique: true,
```
Comment (Alfred-Mountfield) [inline packages/graph/migrations/postgres/migration/1656417312397_initial.ts:33]: Should the `id` be unique? Shouldn't the `version_id` be unique?

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        `),
+    },
+  );
+
+  pgm.createTable(
+    "data_types",
+    {
+      version_id: {
+        type: "UUID",
+        primaryKey: true,
```
Comment (Alfred-Mountfield) [inline packages/graph/migrations/postgres/migration/1656417312397_initial.ts:50]: Are we not referencing `ids` yet then?

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts (hunk 2 above)
Comment (benwerner01) [reply packages/graph/migrations/postgres/migration/1656417312397_initial.ts]: > Should the id be unique?
Good point, more than one `version_id` can have the same `id` - have addressed in https://github.com/hashintel/hash/pull/712/commits/692832839cea317a0ad4ecf08ecb6ed97343adef

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts (hunk 3 above)
Comment (benwerner01) [reply packages/graph/migrations/postgres/migration/1656417312397_initial.ts:49]: Good spot, have addressed in https://github.com/hashintel/hash/pull/712/commits

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts (hunk 2 above)
Comment (Alfred-Mountfield) [reply packages/graph/migrations/postgres/migration/1656417312397_initial.ts]: Gotcha, my lack of knowledge shining through :D

**Review by Alfred-Mountfield — COMMENTED** [review]
LGTM 🚀 
Will approve when the other PRs are merged disabling some of CI and such

File: packages/graph/hash_graph/lib/graph/Cargo.toml [hunk 4]
```diff
@@ -13,6 +13,12 @@ repository.workspace = true
 
 
 [dependencies]
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+
+# We probably don't need full.
+tokio = { version = "1", features = ["full"] }
```
[the remark this answers is not in the document]
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/Cargo.toml]: Addressed in https://github.com/hashintel/hash/pull/712/commits/aa9fb137ae5579c1b7edf880f6165409d09d5de5

File: packages/graph/hash_graph/lib/graph/src/db/mod.rs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+mod tests {
+    use super::row_types::EntityType;
+
+    // Sanity check test, not useful at all. Delete as soon as it makes sense.
+    #[ignore]
+    #[tokio::test]
+    async fn db_works() {
+        let pool = sqlx::PgPool::connect("postgres://postgres:postgres@localhost/postgres_graph")
+            .await
+            .expect("DB to be up and available");
```
[the remark this answers is not in the document]
Comment (benwerner01) [reply packages/graph/hash_graph/lib/graph/src/db/mod.rs]: addressed in https://github.com/hashintel/hash/pull/712/commits/ad93deb465f690dd81113d9f5e5c36be24b4181e

**Review by Alfred-Mountfield — DISMISSED** [review]
🚀

**Review by Alfred-Mountfield — APPROVED** [review]

**Pull request merged**